src/dmqproto/client/request/internal/Consume.d

```diff
@@ -12,6 +12,711 @@
+            // the remaining records. To avoid a dangling slice if
+            // *this.batch_buffer is relocated, set this.remaining_batch to null
+            // first.
+            size_t n_processed = (*this.batch_buffer).length - this.remaining_batch.length;
+            this.remaining_batch = null;
+            (*this.batch_buffer) ~= record_batch;
+            this.remaining_batch = (*this.batch_buffer)[n_processed .. $];
+
+            if (this.fiber_suspended == fiber_suspended.WaitingForRecords)
+                this.fiber.resume(token, this);
+        }
+
+        /***********************************************************************
+
+            Resumes passing records to the user. Called when the user resumes
+            the request through the controller.
+
+        ***********************************************************************/
+
+        public void resume ( )
+        {
+            if (this.fiber_suspended == fiber_suspended.RequestSuspended)
+                this.fiber.resume(token, this);
+        }
+
+        /***********************************************************************
+
+            Requests the fiber to terminate when all remaining records have been
+            passed to the user. Called when a `Stopped` message from the node
+            has arrived.
+
+        ***********************************************************************/
+
+        public void stop ( )
+        {
+            this.stopped = true;
+            if (this.fiber_suspended == fiber_suspended.WaitingForRecords)
+                this.fiber.resume(token, this);
+        }
+
+        /***********************************************************************
+
+            Waits for a batch of records to arrive and passes it to the user,
+            then sends the `Continue` message to the node, in a loop. Handles
+            suspending the request through the controller, for resuming call
+            `resume`. Calling `stop` makes this routine terminate after all
+            remaining records have been passed to the user.
+
+        ***********************************************************************/
+
+        private void fiberMethod ( )
+        {
+            while (this.waitForRecords())
+            {
+                for (uint yield_count = 1; ; yield_count++)
+                {
+                    if (this.outer.context.shared_working.suspendable_control.suspended)
+                    {
+                        yield_count = 0;
+                        this.fiber_suspended = fiber_suspended.RequestSuspended;
+                        this.fiber.suspend(token, this);
+                        this.fiber_suspended = fiber_suspended.No;
+                    }
+
+                    this.passRecordToUser(
+                        this.outer.conn.message_parser.getArray
+                        !(Const!(void))(this.remaining_batch)
+                    );
+
+                    if (!this.remaining_batch.length)
```

Comment:
Just a small point, but it feels to me like it'd be clearer to replace the for loop with (something like) `while ( records_in_batch )`. That's the main point of this loop. The yielding is only a side-issue. As it stands now, it seems the other way around.

Reply:
Can we settle for `for (uint yield_count = 0; this.remaining_batch.length; yield_count++)`?

Reply:
You really think the yielding makes more sense as the outer loop? That seems backwards to me.

Reply:
It is the same as `while (this.remaining_batch.length)` plus a loop-local yield counter.

Reply:
Ahh I see. Sorry, I missed the loop condition.

Seems ok. I still find it a bit harder to read what's happening, but I don't think it's important enough to make an issue out of :)